### PR TITLE
Improve type error messages and tooltips

### PR DESCRIPTION
**Pull request by hasparus**
Okay this might be a small thing, but I want to add some polish (lol) to the types.

Generally, the problem we have is that our types, mostly imported from DT, may be a bit confusing.
I'd like to make Theme UI types as welcoming as possible for v1 (#832)

**Related issues:**

- https://github.com/system-ui/theme-ui/issues/834#issuecomment-625618512
  - [ ] Add TypeScript section to the docs to explain most common tricky things people can see
- https://github.com/system-ui/theme-ui/issues/834#issuecomment-625703446
  - [ ] Move all tests to TypeScript. We can't be sure Theme UI works in TypeScript if we don't typecheck our tests.
  - [x] Get rid of generic parameter for `merge`. Use `Theme` as argument.
- https://github.com/system-ui/theme-ui/issues/907
  - [ ] Get inspired by Emotion's `Interpolation` and display errors on `SystemStyleObject` better.
- We don't support vendor properties. **Why?** They are defined in `csstype` and we actually have them inside `StandardCSSProperties`, but they don't come to `sx` prop. I can't see a reason we shouldn't support vendor properties on type level, while they work at runtime.
  - [x] Let's add `CSS.VendorProperties<number | string>` to `CSSProperties` type.
- [ ] Add `label?: string` to `ThemeUIStyleObject` (https://github.com/DefinitelyTyped/DefinitelyTyped/issues/45020#issuecomment-633300398)
- [x] Rename `SystemStyleObject` to `ThemeUIStyleObject`.
  - System is a loaded word in our industry, and the name is inherited from styled-system. It might be more user-friendly to tidy up few types and use a mix of [nice meaningful names](https://jxnblk.com/blog/design-graph/) and dumb straightforwad names.
  
---

Few notes to explain what I'm doing.

- Interfaces can be merged with interfaces of the same name.
  - https://www.typescriptlang.org/docs/handbook/declaration-merging.html#merging-interfaces
- Type aliases expand in error messages and tooltips.
- Interfaces don't allow property name conflicts in extended types.
  Intersections (`&`) turn a conflicting prop into `never`. This isn't something we often want.
  [(TS Playground with example)](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UDeUAeAuKByAhlqAvgNwBQokUAQgsmplgEb7EklIlRTnQzWJwAyKqU7dYlABo1eAbSyosAXSgB6FVAA8AWigA7CADcIAJxIFW7TmqgBjAPa6AZgBsAljeABCW3YC2YV2cILWBXX2gTYztTTlddYBNHHBseaghUBN0AEwBnWAAaKmRzcyA)

### CODE REVIEW

**Comment by hasparus** [comment]
@jxnblk Which colors in `theme.colors` should be required? ATM it's `primary`, `background` and `text`.

File: packages/theme-ui/test/color-modes.tsx [hunk 1]
```diff
@@ -373,14 +372,14 @@ test('warns when initialColorModeName matches a key in theme.colors.modes', () =
 })
 
 test('dot notation works with color modes', () => {
-  const Button = (props) => {
```
Comment (hasparus) [inline packages/theme-ui/test/color-modes.tsx:376]: damn, can't Prettier decide on one style? I'll revert this change.

**Comment by hasparus** [comment]
This is starting to become pretty big, and while I didn't tackle all todos yet, it might be good to start reviewing before it becomes overwhelming.

**Comment by jxnblk** [comment]
> Which colors in theme.colors should be required?

Generally speaking, none should be *required,* but the color mode functionality does depend on the `colors.text` and `colors.background` keys being defined.

For the interoperability part, I think we'll want to create a "blessed" Theme interface (along with some color palette interfaces) after the TS conversion is out as an optional piece of the puzzle, but not sure if that's helpful here

File: packages/core/src/types.ts [hunk 2]
```diff
@@ -1,11 +1,11 @@
-import { SystemStyleObject } from '@theme-ui/css'
+import { ThemeUIStyleObject } from '@theme-ui/css'
```
Comment (jxnblk) [inline packages/core/src/types.ts:1]: Not to bikeshed, but this is a *much* better name :)

**Pull request merged**